### ISSUE TITLE
Migrate from JUnit 4 to JUnit 5 

### ISSUE THREAD

**Issue by DomGarguilo** [apache/accumulo#2441]
It would be nice to upgrade the project from the older JUnit 4 to the newer JUnit 5 Jupiter dependency. The migration process would be large so I thought it would be a good idea to have a ticket to track the progress of the migration.

It has been suggested that this task be broken up into various chunks so that pull requests are smaller, more manageable and easier to review.

There are a few general steps that this migration can be broken down into:

> 1. things that can be done while still on JUnit4 to ease the migration
> 2. things to update the parent POM to manage the JUnit dependencies
> 3. (module migration stage 1) things that can be done in each module to move that module from JUnit4 to JUnit5 with the vintage engine
> 4. (module migration stage 2) things that can be done in each module to move that module from JUnit5 vintage to JUnit5 Jupiter engine

_Originally posted by @ctubbsii in https://github.com/apache/accumulo/issues/2427#issuecomment-1022535919_

Step 1 from the list above includes things that overlap between JUnit 4 and 5. An example of this sort of change is #2435. At this point I'm not sure if there are any more tasks that fit into step 1.

Step 2 may only require one PR as well.

The following table (_Originally posted by @ctubbsii in https://github.com/apache/accumulo/issues/2427#issuecomment-1022535919_) may aide in keeping track of stages 3 and 4. This table can be updated to reflect the progress of the migration as it happens, using :heavy_check_mark: or something similar to mark completion.
| Module      | Migrate to JUnit5 vintage | Migrate to JUnit5 jupiter |
| ----------- | ----------- | ----------- |
| assemble |  |  |
| core |  |  |
| hadoop-mapreduce |  |  |
| iterator-test-harness |  |  |
| minicluster |  |  |
| server/base |  |  |
| server/compaction-coordinator |  |  |
| server/compactor |  |  |
| server/gc |  |  |
| server/manager |  |  |
| server/master |  |  |
| server/monitor |  |  |
| server/native |  |  |
| server/tserver |  |  |
| shell |  |  |
| start | |  |
| test |  |  |

**Comment by DomGarguilo** [comment]
Note: I don't think this is a priority at the moment and most changes should wait to me merged until after the 2.1.0 release so I will tentatively label this ticket with the 3.0 project tag.

**Comment by DomGarguilo** [comment]
In JUnit 5, the assert statements can accept a `Supplier<String>` as an alternative to `String` for the message to supply if the assert fails. For example:
```java
    Supplier<String> message = () -> "Calculator.add(2, 2) test failed";
    assertEquals(3, Calculator.add(2, 2), message);
```
or equivalently 
```java
    assertEquals(3, Calculator.add(2, 2), () -> "Calculator.add(2, 2) test failed");
```
This has the benefit of being lazily evaluated so the string will only be created when the message is needed rather than always being created. I am wondering if this possible performance gain is worth the possible (small) decrease in readability/increase in code complexity. This could be something we standardize while the migration of JUnit 4 to 5 takes place since we need to change the order of the assert parameters when there is a message anyways.


Anyone have any thoughts on this?

**Comment by AlbertWhitlock** [comment]
I like this way of doing it. The message will only get created when needed and I don't think the code becomes much less readable.

**Comment by ctubbsii** [comment]
The vast vast vast majority of JUnit assert statements are trivial and obvious, and don't need a special message at all, because the message that JUnit itself provides is more than sufficient. Even if the extra time for each string took 100ms per test (which I think is an overestimate), that's only a single extra minute on a more than 4 hour build. Even if they took half a second each, that's still only 5 minutes extra on a more than 4 hour build.

Don't bother. It's almost certainly not worth it.

**Comment by EdColeman** [comment]
It is a test so readability and maintainability should have priority.  There is no way that any optimization of a string that is executed once during a test is going to make any difference.  Something like needing repeatedly spin up a minicluser needs to be evaluated. The time to build a log message is not something to worry about.

If it makes migration easier - pick whatever is easiest to understand in terms of the test context.

**Comment by DomGarguilo** [comment]
Thanks for feedback. I wont bother using the supplier. Seems like plain old `String` is sufficient.

**Comment by DomGarguilo** [comment]
Here are some quick notes that might help to reference for those who work on the migration or for reviewers.

| JUnit4  | JUnit5  |
| ----------- | ----------- |
| `@Test` | `@Test` |
| `@Before` | `@BeforeEach` |
| `@After` | `@AfterEach` |
| `@BeforeClass` | `@BeforeAll` |
| `@AfterClass` | `AfterAll` |
| `@Ignore` | `@Disable` |
| `@Category` | `@Tag` |

| | JUnit4  | JUnit5  |
| ----------- | ----------- | ----------- |
| Testing class package | `org.junit` | `org.junit.jupiter.api` |
| Assertions class | `Assert` | `Assertions` |

`TemporaryFolder` was removed and instead a `File` or `Path` is used in JUnit5 with the `@TempDir` annotation.

**Comment by DomGarguilo** [comment]
I wanted to bring up another case that could be standardized while converting to JUnit5. The new `@Timeout` annotation which can be specified per class or per method has a default unit of seconds (e.g. `@Timeout(42)` would be 42 seconds). There is an optional parameter that can be added to specify units (e.g. `@Timeout(value = 42, unit = MINUTES)` would be 42 minutes). [See the docs for more details](https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/Timeout.html).

Currently, the ITs specify a timeout in seconds, usually doing something like `5 * 60` to specify minutes (5 mins in this case).

My question is, should we keep everything standard and continue using seconds for all timeouts or, when it makes sense, use the unit parameter (e.g. to specify a 6 minute timeout, instead of `@Timeout(5 * 60)` should we use `@Timeout(5, unit = MINUTES)`).

This is pretty small but there are a lot of occurrences so I thought I would get some opinions before I convert them all.

**Comment by ctubbsii** [comment]
I don't think it matters much. If you want to specify SECONDS explicitly, it's fine. Otherwise, I think it's pretty easy to infer SECONDS unless otherwise specified.

**Comment by DomGarguilo** [comment]
I'm running into some issues getting the timeouts configured with JUnit 5 `@Timeout`. Currently, for ITs, there is a default timeout that is set in `AccumuloITBase` and then can be altered with a timeout factor that can be supplied as an argument when running the tests in terminal. All tests that extend `AccumuloITBase` have this timeout applied to them via the `@Rule`. The code for this is here:
https://github.com/apache/accumulo/blob/f8ec16c833ad7b58759a9101e4d9befd524b9daf/test/src/main/java/org/apache/accumulo/harness/AccumuloITBase.java#L101-L122

The `Timeout` object and `@Rule` are no longer present in JUnit5. As far as I can tell, there are ways to set a global timeout to tests(via [Configuration Parameters](https://junit.org/junit5/docs/current/user-guide/#running-tests-config-params)), but the issue I'm running into is how to incorporate the timeout factor that is used to scale the timeouts. The `@Timeout` annotations need to be set per class and will not apply to child classes. The value for `@Timeout`s can be set to a variable but it needs to be a constant meaning we cannot set the timeout to something like `final static int TIMEOUT = 600 * getTimeoutFactor()`.

**Comment by DomGarguilo** [comment]
In regards to my previous comment, a ticket has been filed in JUnit5 upstream to inquire about this issue: https://github.com/junit-team/junit5/issues/2841

**Comment by jbduncan** [comment]
I don't know if you're all aware, but there are a number of tools out there that automate some of the work on translating JUnit 4 tests to JUnit 5 ones:

- https://github.com/sbrannen/junit-converters (by one of the JUnit 5 maintainers)
- https://github.com/junit-pioneer/convert-junit4-to-junit5
- https://github.com/openrewrite/rewrite-testing-frameworks (to be used with [OpenRewrite](https://github.com/openrewrite/rewrite)).

**Comment by Devamparikh** [comment]
Hey there, Can outsiders work on this issue? I want to help in migrating Junit4 to Junit5. Can I get some issues to work on? Thanks :)

**Comment by Devamparikh** [comment]
Actually I have applied for GSoC2022 and asked to work on  Eclipse JKube™: Migrate To JUnit5 Testing Framework from JUnit4 project. I would appreciate it if you could assign a small portion of the migration to JUnit5 to me, please. And I assure you I'll work hard.

**Comment by ctubbsii** [comment]
@Devamparikh We appreciate your interest, but this work is already being done by others and is almost done, and there isn't anything left to share with you. If there's other tickets you're interested in working on, we would welcome your help with those instead. But, unfortunately, I don't think there's anything left for you to help with on this one.

**Comment by Devamparikh** [comment]
@ctubbsii Okay, Let me check In what ways can I be helpful.

**Comment by DomGarguilo** [comment]
I think everything that can be done at this point has been done for this ticket. The only JUnit4 tests that remain are those that use powermock. Im wondering if this ticket should stay open until we have completely removed all JUnit4 usage or if it should be closed and the remaining tests that use powermock be tracked in a new ticket.

**Comment by ctubbsii** [comment]
> I think everything that can be done at this point has been done for this ticket. The only JUnit4 tests that remain are those that use powermock. Im wondering if this ticket should stay open until we have completely removed all JUnit4 usage or if it should be closed and the remaining tests that use powermock be tracked in a new ticket.

The only remaining thing I can think of for this issue is to add a check that:

1. Fails if it detects uses of the vintage API in classes other than the few that currently need to use it for PowerMock.
2. Fails if it detects that any jupiter API being used in one of those same few classes.

After that, we should be able to close it as complete. There's nothing to be done for the few PowerMock classes, so there's nothing else to track in a new ticket.

**Comment by ctubbsii** [comment]
I didn't realize there was already an open ticket for this: #2586 , so this can be closed now.
